### PR TITLE
Implement ISOBUS File Server (FS) Interface as a Personal Project

### PR DESCRIPTION
**Pull request by olerem**
Introduce the ISOBUS File Server (FS) interface, compatible with ISO 11783-13. The implementation utilizes the kernel's existing CAN J1939 socket support.

### CODE REVIEW

File: isobusfs/isobusfs_srv_fa.c [hunk 1]
```diff
@@ -0,0 +1,923 @@
...
+	return ret;
+
+not_supported:
+	isobusfs_srv_send_error(priv, msg, ISOBUSFS_ERR_FUNC_NOT_SUPPORTED);
+
+	pr_warn("%s: unsupported function: %i", __func__, func);
+
+	/* Not a critical error */
+	return 0;
+}
```
Comment (vincent-mailhol) [inline isobusfs/isobusfs_srv_fa.c:923]: Simplify as:
```suggestion
	default:
		isobusfs_srv_send_error(priv, msg, ISOBUSFS_ERR_FUNC_NOT_SUPPORTED);

		pr_warn("%s: unsupported function: %i", __func__, func);

		/* Not a critical error */
	}

	return 0;
}
```
then remove `ret` declaration.

File: isobusfs/isobusfs_srv_fh.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	int ret = 0;
+
+	switch (func) {
+	case ISOBUSFS_FH_F_MOVE_FILE_REQ:
+	case ISOBUSFS_FH_F_DELETE_FILE_REQ:
+	case ISOBUSFS_FH_F_GET_FILE_ATTR_REQ:
+	case ISOBUSFS_FH_F_SET_FILE_ATTR_REQ:
+	case ISOBUSFS_FH_F_GET_FILE_DATETIME_REQ:
+	default:
+		goto not_supported;
```
Comment (vincent-mailhol) [inline isobusfs/isobusfs_srv_fh.c:24]: Same as above.

File: isobusfs/isobusfs_srv_vh.c [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+/* Command group: volume hnadling */
+int isobusfs_srv_rx_cg_vh(struct isobusfs_srv_priv *priv,
+			  struct isobusfs_msg *msg)
+{
+	int func = isobusfs_buf_to_function(msg->buf);
+	int ret = 0;
+
+	switch (func) {
+	case ISOBUSFS_VA_F_INITIALIZE_VOLUME_REQ: /* fall through */
+	default:
```
Comment (vincent-mailhol) [inline isobusfs/isobusfs_srv_vh.c:19]: Never supported? If this is for future development, maybe add a `TODO` comment?

File: isobusfs/isobusfs_cli.c [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+	if (!found)
+		return -ENOENT;
+
+	for (; i < priv->num_events - 1; ++i)
+		priv->events[i] = priv->events[i + 1];
+
+	priv->num_events--;
+
+	return 0;
+}
```
Comment (vincent-mailhol) [inline isobusfs/isobusfs_cli.c:66]: Maybe:
```suggestion
	for (i = 0; i < priv->num_events; ++i) {
		if (&priv->events[i] == event_to_remove) {
			memmove(&priv->events[i], &priv->events[i + 1],
				(priv->num_events - i - 1) * sizeof(*priv->events));
			priv->num_events--;
			return 0;
		}
	}

	return -ENOENT;
}
```

**Review by vincent-mailhol — COMMENTED** [review]
Not a real review but more a quick scroll through. I never used Open ISOBUS and do not know the protocol, so these are just uneducated comments.

Overall, it seems that a lot of efforts were put here. The codes look well structure, well commented and well tested. I will unfortunately not be able to put more effort in this review, so just wish you the best on your project!

**Comment by oak-tree** [comment]
this is very interesting project, any plan to merge it in the near future?

**Comment by olerem** [comment]
> this is very interesting project, any plan to merge it in the near future?

Sure, i'm working on applying the review comment :)
Can you please tell more on why this project is interesting for you and what kind of use cases do you have?

**Comment by oak-tree** [comment]
[HHey @olerem 
> Sure, i'm working on applying the review comment :) Can you please tell more on why this project is interesting for you and what kind of use cases do you have?
Hey, yes, i missed you question.  For special agriculture uses. both R&D and productions projects.

**Comment by olerem** [comment]
Hey @oak-tree ,

 > Hey, yes, i missed you question.  For special agriculture uses. both R&D and productions projects.

Great. Do you have any way to validate my implementation against real devices?

Do you have some user case stories to promote this project?

**Pull request merged**